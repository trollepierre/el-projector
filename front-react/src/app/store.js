import { withReducer } from '../store/withReducer';
import { apiService, tokenService } from '../services';

const SET_IS_AUTHENTICATED = 'app/SET_IS_AUTHENTICATED';

const setIsAuthenticated = isAuthenticated => dispatch => {
  if(!isAuthenticated){
    tokenService.removeTokens()
  }
  return dispatch({
    type: SET_IS_AUTHENTICATED, payload: { isAuthenticated }
  });
};

const authenticate = password => async dispatch => {
  try {
    const data = {
      email: 'some@example.org',
      name: password
    };
    const tokens = await apiService.post('login', data)
    await tokenService.saveUserTokens(tokens, data)
    dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: true }})
  } catch (error) {
    dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: false }})
    alert('wrong password')
  }
}

const loginSilently = () => async dispatch => {
  try {
    const refreshToken = tokenService.getRefreshToken();
    if(!refreshToken) return;
    const tokens = await apiService.post('login/token', { refreshToken })
    await tokenService.reauthenticate(tokens);
    await dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: true }})
  } catch (error) {
    await dispatch({ type: SET_IS_AUTHENTICATED, payload: { isAuthenticated: false }})
    await tokenService.removeTokens()
    alert('problem during login silently')
    throw error
  }
}

export const actions = {
  setIsAuthenticated,
  authenticate,
  loginSilently,
};

export const initialState = {
  isAuthenticated: false,
};

export const actionHandlers = {
  [SET_IS_AUTHENTICATED]: (state, action) => {
    const { isAuthenticated } = action.payload;
    return {
      ...state,
      isAuthenticated,
    };
  },
};

export const reducer = withReducer(initialState, actionHandlers);
