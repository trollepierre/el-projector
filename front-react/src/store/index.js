export { default as withReducer } from './withReducer'
export { default as createStore } from './createStore'
