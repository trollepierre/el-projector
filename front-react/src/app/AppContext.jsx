import React, { useReducer, createContext, useContext } from 'react';
import PropTypes from 'prop-types';
import { reducer, initialState, actions } from './store';

const AppContext = createContext();

export const useAppContext = () => useContext(AppContext);

const getProps = (state, dispatch) => ({
  ...state,
  setIsAuthenticated: isAuthenticated => actions.setIsAuthenticated(isAuthenticated)(dispatch),
  authenticate: value => actions.authenticate(value)(dispatch),
});

export const AppProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <AppContext.Provider value={{ ...getProps(state, dispatch) }}>
      {children}
    </AppContext.Provider>
  );
};

AppProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default AppContext;
