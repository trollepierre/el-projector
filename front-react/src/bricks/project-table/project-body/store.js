import apiService from '../../../services/api/api-service'
import { withReducer } from '../../../store/withReducer';

export const FETCH_TASKS_STARTED = 'FETCH_TASKS_STARTED';
export const FETCH_TASKS_SUCCEEDED = 'FETCH_TASKS_SUCCEEDED';
export const FETCH_TASKS_FAILED = 'FETCH_TASKS_FAILED';

// this.sortedTasks = apiService.get('tasks').then(tasks => {
//   console.log('tasks');
//   console.log(tasks);
//   return tasks
// })

export const fetchTasks = () => async dispatch => {
  dispatch({ type: FETCH_TASKS_STARTED });

  try {
    // const tasksFixtures = [
    //   {
    //     id: 1,
    //     image_src:
    //       'https://cdn.manomano.fr/martillo-dewalt-dch273n-18v-P-2075566-4162147_1.jpg',
    //     title: 'Demo react Product',
    //     price_currency: '€',
    //     price: '2490.50',
    //   },
    //   {
    //     id: 2,
    //     image_src:
    //       'https://cdn.manomano.com/futbolin-de-acero-60-kg-140x745x875-cm-negro-P-334554-10088120_1.jpg',
    //     title: 'DIY Essential',
    //     price_currency: '€',
    //     price: '240.50',
    //   },
    // ];
    const data = await apiService.get('tasks');
    dispatch({
      type: FETCH_TASKS_SUCCEEDED,
      payload: data,
    });
  } catch (error) {
    dispatch({
      type: FETCH_TASKS_FAILED,
      error,
    });
  }
};

export const initialState = {
  tasks: {
    data: [],
    isLoading: false,
    error: undefined,
  },
};

export const actionHandlers = {
  [FETCH_TASKS_STARTED]: state => {
    return {
      ...state,
      tasks: {
        ...state.tasks,
        error: undefined,
        isLoading: true,
      },
    };
  },
  [FETCH_TASKS_SUCCEEDED]: (state, action) => {
    return {
      ...state,
      tasks: {
        data: action.payload,
        error: undefined,
        isLoading: false,
      },
    };
  },
  [FETCH_TASKS_FAILED]: (state, action) => {
    return {
      ...state,
      tasks: {
        ...state.tasks,
        error: action.error,
        isLoading: false,
      },
    };
  },
};

export const reducer = withReducer(initialState, actionHandlers);

export const tasksSelector = state => state.tasks;

export default reducer;

