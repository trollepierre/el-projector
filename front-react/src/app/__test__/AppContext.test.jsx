import React from 'react';
import { shallow } from 'enzyme';
import AppContext, { AppProvider, getProps } from '../AppContext';
import { actions } from '../store';

describe('AppContext', () => {
  describe('default', () => {
    let realCreateContext;
    let mockCreateContext;

    beforeEach(() => {
      realCreateContext = React.createContext;
      mockCreateContext = realCreateContext;
    });

    afterEach(() => {
      React.createContext = realCreateContext;
    });

    it('should get createContext', () => {
      expect(AppContext).toEqual(mockCreateContext());
    });
  });

  describe('AppProvider', () => {
    it('should get createContext api and inject it in the header context', () => {
      // When
      const wrapper = shallow(
        <AppProvider>I am the children</AppProvider>
      );

      // Then
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('getProps', () => {
    const dispatch = jest.fn();

    it('should plug setIsAuthenticated', () => {
      // Given
      const props = getProps({}, dispatch);
      actions.setIsAuthenticated = jest.fn(() => () => {});

      // When
      props.setIsAuthenticated('isAuthenticated');

      // Then
      expect(actions.setIsAuthenticated).toHaveBeenCalledWith(
        'isAuthenticated',
      );
    });

    it('should plug authenticate', () => {
      // Given
      const props = getProps({}, dispatch);
      actions.authenticate = jest.fn(() => () => {});

      // When
      props.authenticate('password');

      // Then
      expect(actions.authenticate).toHaveBeenCalledWith(
        'password',
      );
    });

    it('should plug loginSilently', () => {
      // Give,
      const props = getProps({}, dispatch);
      actions.loginSilently = jest.fn(() => () => {});

      // When
      props.loginSilently();

      // Then
      expect(actions.loginSilently).toHaveBeenCalledWith();
    });
  });
});

