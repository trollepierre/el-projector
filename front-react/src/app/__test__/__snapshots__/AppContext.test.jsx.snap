// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContext AppProvider should get createContext api and inject it in the header context 1`] = `
<ContextProvider
  value={
    Object {
      "authenticate": [Function],
      "isAuthenticated": false,
      "loginSilently": [Function],
      "setIsAuthenticated": [Function],
    }
  }
>
  I am the children
</ContextProvider>
`;
